[STEP1] Ayaan, 제이푸시

corpiii opened the pull request:
@unchain123 
안녕하세요 언체인! 👋
어제 PR을 드리려고 했지만 너무 늦은 밤이 되어서 지금 PR 보내드립니다 ㅎ

## 📖 코드 설명

- **func generateAnswer() -> [Int]**
  - 1~9까지의 숫자를 섞어준 후 앞에서 부터 3개의 요소를 반환하는 함수
  
- **decideBallCount(comparing: [Int], with userNumbers: [Int]) -> (Int, Int)**
  - 볼, 스트라이크를 판단하는 함수
  - 인덱스의 위치와 값이 같다면 스트라이크, 그렇지 않다면 포함 되어 있는 경우 볼 판단
  
- **playBall()**
  - 야구 게임 진행 함수
  - remainCount 만큼 게임 횟수를 진행
  - 스트라이크, 볼 갯수를 비교해서 결과를 출력

## 🤔 궁금한 점
-  generateAnswer() -> [Int]
array의 값 슬라이싱을 통해 3개의 값을 가져와서 반환하려다 보니 arraySlice형이 반환되어서 arraySlice형으로 반환할 지 Int형 array로 변환 후 반환할지 아얀과 깊은 고민끝에 Array 생성자를 통해 Int형 array로 변환해서 반환하기로 했습니다.
arraySlice형으로 반환해서 값을 받을 때 Array 생성자로 덮어주어 반환하는게 좋을지, Array형으로 먼저 덮어주어 반환하는 방법이 좋을지 언체인의 생각도 궁금합니다!

- playBall()
37번째 줄의 방식에 대해서 궁금합니다!
양식과 같이 출력하기 위해서 String으로 매핑해준 후 구분자를 공백으로 사용하여서 하나의 문자열로 합쳐 출력하였습니다. 
언체인이면 어떻게 사용하셨을지도 궁금합니다

- 순서도
순서도를 작성할 때 변수의 네이밍, 자세한 로직을 설명을 했는데 그러다보니 너무 순서도가 복잡해지고 많은 시간이 소요되었습니다. 순서도의 의미를 생각해서 전체적인 흐름을 파악할 정도로만 작성하는게 좋을지, 만약 이렇게 작성한다면 언체인과 같은 리뷰어들이 보시기에 문제가 없는지 궁금합니다

corpiii committed:
docs: 순서도 이미지 추가

corpiii committed:
docs: playBall 순서도 이미지 수정

corpiii committed:
docs: 순서도 파일 추가

corpiii committed:
feat: generateAnswer 함수 정의

generateAnswer은 3개의 임의의 자연수를 가진 Array타입을 반환하는 함수이다.
1부터 9까지 자연수를 담은 Array타입을 정의한 후 해당 변수를 shuffle함수를 활용해 무작위로 정렬한다.
그 후 앞에서 부터 3개의 자연수만 반환한다.

corpiii committed:
feat: decideBallCount 함수 선언

oneStar92 committed:
feat: playBall 함수 선언.

oneStar92 committed:
feat: playBall 함수 선언.

oneStar92 committed:
refactor: Array생성자를 통해 Array<Int>형을 생성하여 반환하는 형태로 변경.

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
 
 import Foundation
```
Foundation 은 어떤 역할을 하는 프레임워크일까요? 여기서 import해주신 이유가 있을까요?

corpiii replied:
공식문서에 따르면 Foundation은 앱 개발에 필요한 필수 데이터 유형, 컬렉션 등의 기본적인 부분을 도와주는 프레임워크 입니다. 이번 프로젝트에서는 사용하지 않는데 기본적으로 적혀있던 것을 지우지 않고 사용했습니다. 제거 했습니다!

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
 import Foundation
 
-print("Hello, World!")
+func generateAnswer() -> [Int] {
```
generateAnswer라는 함수명은 정답을 만든다?라고 해석이 되는데 이 함수는 정답을 만들고 있을까요?? 다른 변수명을 한 번 생각해 볼까요??

oneStar92 replied:
fetchRandomNumbers으로 변경하였습니다!

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
+    var array = Array(1...9)
+    array.shuffle()
+    return Array(array[...2])
+}
```
shuffle을 사용해주셨네요! 이 방법도 좋은 방법 인것 같습니다!
하지만 프로젝트 기능요구서에서 아래사진에 나온 경험을 프로젝트 수행중 얻을 수 있는 핵심경험이라고 나와있습니다.  기능 구현도 좋지만 프로젝트에서 할 수 있는 경험들을 최대한 경험해 보는 것도 좋을 것 같습니다.
<img src=https://i.imgur.com/8CDa9K8.png width=250>

oneStar92 replied:
요구사항을 제대로 확인하지 못한 것 같습니다!
요구사항을 반영하여 수정하였습니다.

unchain123 replied:
무조건 바꾸라는 건 아니었지만 바꿔 보시면서 Set에 대해서도 이해 할 수 있는 시간이 됐을 것 같아요 🥳

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
+    return Array(array[...2])
+}
 
+func decideBallCount(comparing answer: [Int], with userNumbers: [Int]) -> (Int, Int) {
```
이 메서드명은 BallCount만을 결정하는 메서드인것 처럼 보이기도 합니다. 좀 더 명확한 네이밍을 해 볼 수 있을까요?

oneStar92 replied:
야구에서 Strike와 Ball의 카운트를 합쳐서 Ball 카운트라고 불러서 해당 부분을 decideBallCount라고 네이밍하였었습니다.
하지만 다시 한번더 고민해 보니 의미가 애매한 것 같기에 decideStrikeCountAndBallCount로 네이밍을 하였습니다.

unchain123 replied:
메서드명 변경하시느라 고생하셨습니다. 저희도 아직까지 고민하는 부분이에요.
`decideStrikeCountAndBallCount` 이 메서드는 count가 중복되고 decide라는 단어가 뭔가 결정한다는 의미를 담고 있어서 어색해 보입니다. 메서드명을 만들 때 동일한 단어가 중복되는건 피하는게 좋습니다. 이 부분은 다음 스탭에서 반영해주세요

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
+    for index in 0...2 {
+        if userNumbers[index] == answer[index] {
+            strikeCount += 1
+        } else if answer.contains(userNumbers[index]) {
+            ballCount += 1
```
contains를 활용해서 ball카운트를 확인해 보는 부분이 아주 좋습니다!

corpiii replied:
감사합니다! 😁

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
+    return (strikeCount, ballCount)
+}
+
+func playBall() {
```
이 메서드가 너무 많은 일을 하고 있는 것 같아요 조금 분리 해 볼 수 있을까요??
그리고 playBall이라는 네이밍은 이 프로그램에서는 조금 어색한것 같습니다. 다른 네이밍을 생각해 볼까요?

corpiii replied:
확실히 조금 애매했던 부분인 것 같아서 startNumberGame 으로 네이밍 변경 하였습니다.

내부의 기능들도 각각
- printInformation(of userNumbers: [Int])
  - 임의의 수를 출력하는 함수
- printDecision(about strikeCount: Int, and ballCount: Int)
  - 스트라이크, 볼의 갯수를 출력하는 함수
- isStrikeOut(_ strikeCount: Int) -> Bool
  - 스트라이크 아웃인지 판단하는 함수
- isGameOver() -> Bool
  - 남은 기회가 0인지 확인하는 함수

로 분리했습니다!

unchain123 replied:
기능 분리가 잘 된건 같아서 보기 좋습니다 😆

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
+
+func playBall() {
+    let answer = generateAnswer()
+    var remainCount = 9
```
프로젝트 요구서에 보시면 아래와 같이 두개의 전역변수를 만들어 달라고 했는데 지역변수로 해주신 이유가 있을까요?
<img src=https://i.imgur.com/4LWlwaf.png width=250>

oneStar92 replied:
이것도 요구사항을 제대로 확인하지 못했네요...
수정하였습니다.

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
+        if strikeCount == 3 {
+            guidance += "사용자 승리!"
+            print(guidance)
+            break;
```
`break;` 라고 쓰는 방식은 C언어에서 사용하는 방법이라고 알고 있습니다. 스위프트에서는 ;을 사용하지 않습니다.

oneStar92 replied:
앗.. 이런 부분이 존재했네요. 수정하였습니다.

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
+    var userNumbers = [Int]()
+    
+    while remainCount > 0 {
+        var guidance = ""
```
guidance라는 변수명은 어떤걸 나타내 주는 변수인가요? 변수명이 조금 어색한 것 같아요

oneStar92 replied:
안내 문구를 출력한다는 의미로 사용하였던 변수명입니다.
하지만 현재 출력부분들 대부분을 함수를 통해서 print해주는 방식으로 변경하여 해당 부분은 이제 사용하지 않습니다.

unchain123 replied:
넵 좋습니다😎

unchain123 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,54 @@
+    while remainCount > 0 {
+        var guidance = ""
+        userNumbers = generateAnswer()
+        let numbersToString = userNumbers.map({String($0)}).joined(separator: " ")
```
고차함수를 사용해 주셨네요! 고차함수를 사용하는 것은 코드를 간결하게 해주는 장점이 있습니다. 하지만 배우는 입장이니 아직 고차함수의 축약형 보다는 기본형을 한번 공부해 보는 것도 좋을 것 같아요. (추후에 고차함수를 활용할 프로젝트가 존재합니다.)

numbersToString변수명이 좀 어색한것 같아요. 다른 변수명을 생각해볼까요?😃

oneStar92 replied:
언체인의 리뷰를 보고 고차함수를 풀어서 써보니 생각보다 잘 써지지가 않더군요. 아직 축약형보단 기본형을 더 익히는데 중점을 두겠습니다.
numbersToString 부분은 printInformation 함수 부분으로 옮겨지면서 네이밍을 다시 하였습니다.

unchain123 replied:
네네 고차함수 기본형에 익숙해지면 고차함수의 사용법도 익숙해지고 원리를 더 잘 이해할 수 있습니다.😍

unchain123 reviewed (commented):
첫프로젝트의 STEP1PR 너무 잘해 주셨습니다. 전체적으로 컨벤션도 잘 지켜진 것 같고 로직도 잘 짜주셨습니다. 다만 조금 어색한 변수명들이 보여서 멘트를 남겼습니다. 코드에는 정답이 없기 때문에 제가 말하는 부분이 절대 정답은 아닙니다. 다만 먼저 이 프로젝트를 지나온 사람으로써 조언을 드리는 거라고 생각해 주시면 됩니다.  제가 멘트를 남긴 부분에서도 두분이 이 코드를 작성한 이유가 명확하다면 그 부분에 대해서도 알려주시면 좋겠습니다. 😎

## 질문에 대한 답변

generateAnswer() -> [Int]
array의 값 슬라이싱을 통해 3개의 값을 가져와서 반환하려다 보니 arraySlice형이 반환되어서 arraySlice형으로 반환할 지 Int형 array로 변환 후 반환할지 아얀과 깊은 고민끝에 Array 생성자를 통해 Int형 array로 변환해서 반환하기로 했습니다.
arraySlice형으로 반환해서 값을 받을 때 Array 생성자로 덮어주어 반환하는게 좋을지, Array형으로 먼저 덮어주어 반환하는 방법이 좋을지 언체인의 생각도 궁금합니다!

- 제각 숫자야구 프로젝트를 할때는 아얀과 제이푸시가 사용한 방법과는 달랐기 때문에 이번에 arraySlice타입을 처음 찾아봤습니다. 그래서 먼저 질문에 답변을 드리자면 return 값으로는 상수에 담아줘서 상수명을 통해 리턴하는 값이 어떤 값인지를 보여주는 것이 좋다고 생각합니다. 

- 그리고 프로젝트 기능요구서에 보면 ![](https://i.imgur.com/9V5czax.png) 라고 명시 되어 있는데 현재의 방법을 선택하신 이유가 있을까요?

playBall()
37번째 줄의 방식에 대해서 궁금합니다!
양식과 같이 출력하기 위해서 String으로 매핑해준 후 구분자를 공백으로 사용하여서 하나의 문자열로 합쳐 출력하였습니다.
언체인이면 어떻게 사용하셨을지도 궁금합니다

- 어떤 부분이 궁금하신건지 잘 이해하지 못했습니다.
- 제가 사용한 방법이 궁금한 것이라면 저는 유저의 인풋을 출력하기 위한 메서드를 만들어서 문자열 보간법을 사용하여 처리해 주었습니다. 


순서도
순서도를 작성할 때 변수의 네이밍, 자세한 로직을 설명을 했는데 그러다보니 너무 순서도가 복잡해지고 많은 시간이 소요되었습니다. 순서도의 의미를 생각해서 전체적인 흐름을 파악할 정도로만 작성하는게 좋을지, 만약 이렇게 작성한다면 언체인과 같은 리뷰어들이 보시기에 문제가 없는지 궁금합니다

- 순서도를 확인해봤습니다. 상당히 자세하고 깔끔한 순서도라고 생각됩니다. 순서도나 UML은 자세할 수록 좋습니다. 후에 이 프로젝트를 유지보수하게 될 불특정 다수(미래의나)에게 더 쉽게 이 코드를 이해할 수 있도록 도와줄 수 있기 때문입니다. 내가 만든 코드이더라도 시간이 지나면 어떤 흐름을 가지는지 잘 기억나지 않을 때 이 순서도만을 보고도 파악 할 수 있다면 좋겠죠.

- 순서도를 어떻게 그릴지는 프로젝트를 진행하는 두분이서 정하시면 됩니다. 지금처럼 자세해도 되고 말씀하신 것처럼 전체적인 흐름만을 파악하는 정도로 하셔도 됩니다. 코드에 정답이 없듯이 여기에도 정답은 없습니다. 다만 나중에 누군가 봤을 때 이 순서도만 갖고 전체 코드를 파악할 수 있으면 좋겠습니다.

- 순서도에서 하나 질문하고 싶은게 있는데 <- 이런식으로 화살표 방향이 되어 있는게 있던데 이게 어떤 의미인지 알 수 있을까요??

oneStar92 committed:
chore: import Foundation 삭제.

corpiii committed:
refactor: fetchRandomNumbers로 네이밍 변경

oneStar92 committed:
refactor: Set타입을 사용하여 randomNumbers를 생성하는 방식으로 변경.

corpiii committed:
refactor: decideStrikeCountAndBallCount로 네이밍 변경.

oneStar92 committed:
refactor: answer, remainCount를 전역변수로 변경.

answer가 전역변수로 변경되어서 decideStrikeCountAndBallCount함수의 parameter로 answer를 전달받지 않아도 되기에 해당부분의 네이밍을 수정했다.

corpiii committed:
refactor: startNumberGame으로 네이밍 변경.

oneStar92 committed:
feat: printInformation함수 선언.

corpiii committed:
feat: printDecision 함수 선언.

oneStar92 committed:
feat: isStrikeOut함수 선언.

corpiii committed:
feat: isGameOver 함수 선언.

corpiii commented:
그리고 프로젝트 기능요구서에 보면 라고 명시 되어 있는데 현재의 방법을 선택하신 이유가 있을까요?
- 아.. 힌트라고 해서 다른방법을 써도 괜찮을 것 같다고 생각했었습니다. 수정했습니다!

순서도에서 하나 질문하고 싶은게 있는데 <- 이런식으로 화살표 방향이 되어 있는게 있던데 이게 어떤 의미인지 알 수 있을까요??
- 순서도를 처음 배울 때 <- 를 사용한다고 배워서 화살표를 사용했습니다. 다음부터는 = 을 사용하도록 하겠습니다!

unchain123 commented:
전체적으로 제가 말씀 드린부분을 잘 적용해 주신것 같습니다. 
다만 제가 말씀 드리는 부분이 절대 정답이 아니고 코드에는 정답이 없습니다. 여러분들의 의견이 있다면 댓글로 달아주셔도 되구요. 
그리고 기능요구서를 상세히 읽는 것도 매우 좋은 자세입니다. 
다음 스탭이 기대 되네요 고생많으셨습니다 😎
